### DISCUSSION TITLE
Why is Next.js loading error.js, not-found.js, and global-error.js on all my pages?

### DISCUSSION THREAD

**Discussion by lancejpollard** [vercel/next.js#63119] [Help] [answered]
### Summary

Here is what I see during development at least on successful page load:

<img width="1151" alt="Screenshot 2024-03-10 at 11 57 36 PM" src="https://github.com/vercel/next.js/assets/28718/2582e5e2-e9a7-42cf-8642-e53aa2a83022">

Why is it loading the 3 error pages? It is blocking, so it delays my page's rendering.

**Comment by icyJoseph** [comment]
I think there's most definitely something odd happening here.

1.6MB for your main-app, and 1MB for all of the others... that's odd.

That being said, can you see in the HTML if the __next_error, I think it is called like that, has been triggered? It might be that your app is falling back to CSR because of a failure during SSR, which would be why all error boundaries are downloaded.

**Reply by lancejpollard** [reply]
@icyJoseph  The 1MB is because I am accidentally internally importing a bunch of JSON from a library I made with a bunch of config stuff, I need to abstract that out and dynamically import the little pieces I need instead of _everything_. It is also imported in my error page accidentally.

How do I tell if `__next_error` is triggered? I don't see the growl notification on the lower left showing any errors or anything. My terminal says `Compiled <page> in Xms`, so everything seems to be working there, unless I am missing something?

What were you thinking of that I should be seeing or debugging in particular?

My app's pages are structured as follows (example page, `/display/equation`):

**`app/display/equation/page.tsx`** (no `use client`)

```tsx
import DisplayEquationPage from '~/components/app/display/equation/Page'
import { convertModuleToObject } from '~/utilities/object'
import { equationMetadata } from '~/values/metadata'

export const metadata = equationMetadata

export default async function Page() {
  const TOOL = convertModuleToObject(await import('~/values/tool'))
  return <DisplayEquationPage TOOL={TOOL} />
}
````

**`components/display/equation/Page.tsx`** (a component, not a special `app` page, with `use client`):

```tsx
'use client'

export default function Page() {
  return "hello world"
}
```

I needed to use `use client` for the entire page layout essentially, because my layout is based on the viewport width (it is responsive), and so I hide the layout when rendered on the server, and show it when calculated properly on the client. I use `useState` directly in `components/**/Page.tsx` all over the place, as well as `useEffect`.

Thoughts?

**Reply by icyJoseph** [reply]
Alright, then it doesn't look like there was an issue there. 

Without looking at code I wouldn't know what else to suggest. What Next.js version do you have? Is there any amount of change that stops triggering this?

**Reply by lancejpollard** [reply]
I am on next canary, a pretty recent one within the last few weeks:

```
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 23.2.0: Wed Nov 15 21:54:51 PST 2023; root:xnu-10002.61.3~2/RELEASE_ARM64_T6030
  Available memory (MB): 36864
  Available CPU cores: 12
Binaries:
  Node: 20.10.0
  npm: 10.4.0
  Yarn: 1.22.19
  pnpm: 8.15.2
Relevant Packages:
  next: 14.1.2-canary.1 // There is a newer canary version (14.2.0-canary.13) available, please upgrade!
  eslint-config-next: 13.1.5
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.3.3
```

**Reply by lancejpollard** [reply]
@icyJoseph even having this in my page.tsx shows the error pages are still loading:

```tsx
export default async function Page() {
  return 'Hello world'
}
```

That is, the `page.tsx` onl says `hello world` with no other imports.

<img width="1062" alt="Screenshot 2024-03-11 at 6 33 56 AM" src="https://github.com/vercel/next.js/assets/28718/ed126a57-3391-4e56-b767-4fe75e98e2a8">

I guess my layout has some imports. If I make my `layout.tsx` this:

```tsx
// 'use client'

// import 'katex/dist/katex.min.css'
// import '@termsurf/vine/base.css'
// import '~/styles/head.css'
// import '~/styles/code.css'
// // import { Analytics } from '@vercel/analytics/react'
// import Container from '@termsurf/vine/component/Container'
// import { DataStorageContextProvider } from '~/hooks/useDataStorage'
// import { FontProvider } from '~/hooks/useFonts'

export default function AppLayout({
  children,
}: {
  children: React.ReactNode
}) {
  return (
    <html>
      <body>{children}</body>
    </html>
  )
}
```

I still see the error p ages loading.

<img width="838" alt="Screenshot 2024-03-11 at 6 37 03 AM" src="https://github.com/vercel/next.js/assets/28718/455baff3-7189-4f53-8ef1-630b284792bc">

What do you see in the network panel, if I may ask? Do you see the error pages loading js?

**Reply by icyJoseph** [reply]
Not in the projects I have at hand right now. Let me try with a fresh out of the box app.

**Reply by icyJoseph** [reply, marked as answer]
Aha, I think I can make an explanation now.

I have used mostly error boundaries, rather than error pages, and my not-found pages are usually server components.

However, if we think about it, when we slap a `use client` on the boundary, whether it catches not-found or loading, we tell the bundler to ship the code for this page to the client.

Since the global-error, or error, or not-found, wrap the React tree, rendering the page, implies rendering them too, since these are just Error Boundaries. 

I am still a bit surprised though... but I guess, it makes sense.

**Reply by lancejpollard** [reply]
@icyJoseph do you think that the error pages are calculated in the "First Load JS"? Maybe that's why I'm really struggling to get below 400kb? https://stackoverflow.com/questions/78139307/logic-behind-how-next-js-builds-dynamic-pages-using-the-app-router-system

**Reply by icyJoseph** [reply]
mm I wouldn't say calculated, but it does look like the JS to do so is being included. I need to setup a couple of projects and investigate this further.
